samples/apps/tinyRA/tinyra/tui.py

```diff
@@ -0,0 +1,1186 @@
+        )
+        user.register_reply([Agent, None], post_last_assistant_msg_to_chat_history)
+
+        # register tools for assistant and user
+        tools = APP_CONFIG.get_tools()
+        for tool in tools.values():
+            description = tool.description
+            code = tool.code
+            # convert a code string to function and return a callable
+
+            function_name, tool_instance = string_to_function(code)
+            assistant.register_for_llm(name=function_name, description=description)(tool_instance)
+            user.register_for_execution()(tool_instance)
+
+        logging.info("Current history:")
+        logging.info(assistant.chat_messages[user])
+
+        await user.a_initiate_chat(assistant, message=task, clear_history=False)
+
+        await user.a_send(
+            f"""Based on the results in above conversation, create a response for the user.
+While computing the response, remember that this conversation was your inner mono-logue. The user does not need to know every detail of the conversation.
+All they want to see is the appropriate result for their task (repeated below) in a manner that would be most useful.
+The task was: {task}
+
+There is no need to use the word TERMINATE in this response.
+            """,
+            assistant,
+            request_reply=False,
+            silent=True,
+        )
+        response = await assistant.a_generate_reply(assistant.chat_messages[user], user)
+        await assistant.a_send(response, user, request_reply=False, silent=True)
+
+        response = assistant.chat_messages[user][-1]["content"]
+
+        await a_insert_chat_message("assistant", response, root_id=0, id=msg_idx + 1)
+
+
+def run_app() -> None:
+    """
+    Run the TinyRA app.
+    """
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--reset", action="store_true", help="Reset chat history")
+    parser.add_argument("--reset-all", action="store_true", help="Reset chat history and delete data path")
+    args = parser.parse_args()
+
+    if args.reset_all:
+        print(f"Warning: Resetting chat history and deleting data path {APP_CONFIG.get_data_path()}")
+        print("Press enter to continue or Ctrl+C to cancel.")
+        input()
+        if os.path.exists(APP_CONFIG.get_database_path()):
+            os.remove(APP_CONFIG.get_database_path())
+        if os.path.exists(APP_CONFIG.get_data_path()):
+            shutil.rmtree(APP_CONFIG.get_data_path())
+        exit()
+
+    if args.reset:
+        print(f"Warning: Resetting chat history. This will delete all chat history in {APP_CONFIG.get_database_path()}")
+        print("Press enter to continue or Ctrl+C to cancel.")
+        input()
+        if os.path.exists(APP_CONFIG.get_database_path()):
+            os.remove(APP_CONFIG.get_database_path())
+        exit()
+
+    app = TinyRA()
+    app.run()
+
+
+if __name__ == "__main__":
+    run_app()
```

Comment:
At > 1k lines this file is starting to become big.  Consider breaking this up into small files